### PR TITLE
fix for https://github.com/thenewmotion/akka-rabbitmq/issues/27

### PR DESCRIPTION
**Pull request by wpc009**
Add a delay for connection recovery. Which will cause connectionActor handle `ProvideChannel` message is `Disconnected` State. After the connection recovered, then send fresh channels to it's children.

### CODE REVIEW

File: src/test/scala/com/thenewmotion/akka/rabbitmq/ConnectionActorSpec.scala [hunk 1]
```diff
@@ -129,8 +131,9 @@ class ConnectionActorSpec extends ActorSpec with Mockito with NoTimeConversions
 
       // give connection actor the time to close and reconnect
       expectMsg(ParentShutdownSignal)
+      Thread.sleep(3000)
```
Comment (t3hnar) [inline src/test/scala/com/thenewmotion/akka/rabbitmq/ConnectionActorSpec.scala]: I'd rather not use `Thread.sleep` even in tests.

File: src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala [hunk 2]
```diff
@@ -23,7 +23,7 @@ object ConnectionActor {
 
   def props(
     factory: ConnectionFactory,
-    reconnectionDelay: FiniteDuration = 10.seconds,
+    reconnectionDelay: FiniteDuration = 3.seconds,
```
Comment (t3hnar) [inline src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala]: Something I'd left to be unchanged.

File: src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala [hunk 3]
```diff
@@ -124,7 +124,7 @@ class ConnectionActor(
     log.debug("{} closing broken connection {}", self.path, connection)
     closeIfOpen(connection)
 
-    self ! Connect
+    setTimer(reconnectTimer, Connect, reconnectionDelay / 3, repeat = false)
```
Comment (t3hnar) [inline src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala]: I still do not understand on how it is supposed to fix something?

Comment (wpc009) [inline src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala]: I add a small delay before the ConnectionActor try to re-establish the connection. This will cause the `ConnectionActor` stays in `Disconnected` state when it receive the `ProvideChannel` message which sent by children actors acquiring for new channel.
Then when the connection recovered, ConnectionActor will create new channels for each of it's children. 
So, children got new channels, and no duplicated channel created, no conflicts.

Comment (wpc009) [inline src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala]: I tried to fix this using a different way. The cause of this issue is ChannelActor get two channel in one reconnecting event. 
1. The first time is when ChannelActor get a Shutdown signal. it drops the old channel and send ConnectionActor a ProvideChannel message.
2. After the ConnectionActor re-establish the connection, it creates new channels for each of it's children automatically.

This two channel cause additional close operation, and triggers some race condition. interrupt some `subscribe` or `I/O reading` operation.

I haven't succeed to remove one of channel creation logic. It seems each of them is required under different scenario. 

I couldn't fix this under a global view. So, may be you guys have better idea.

File: src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala (hunk 2 above)
Comment (wpc009) [inline src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala]: the default 10 seconds for reconnecting delay is a little bit too long. Do you think?

File: src/test/scala/com/thenewmotion/akka/rabbitmq/ConnectionActorSpec.scala (hunk 1 above)
Comment (wpc009) [inline src/test/scala/com/thenewmotion/akka/rabbitmq/ConnectionActorSpec.scala]: I want to waiting for the reconnecting delay. 
How can I do it correctly?

File: src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala (hunk 2 above)
Comment (t3hnar) [inline src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala]: I can agree with you, however it does not mean it is safe to change such value in open sourced library

File: src/test/scala/com/thenewmotion/akka/rabbitmq/ConnectionActorSpec.scala (hunk 1 above)
Comment (t3hnar) [inline src/test/scala/com/thenewmotion/akka/rabbitmq/ConnectionActorSpec.scala]: something like `expectNoMsg(300ms)`

File: src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala (hunk 2 above)
Comment (wpc009) [inline src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala]: I see.  I should revert this change.

File: src/test/scala/com/thenewmotion/akka/rabbitmq/ConnectionActorSpec.scala (hunk 1 above)
Comment (wpc009) [inline src/test/scala/com/thenewmotion/akka/rabbitmq/ConnectionActorSpec.scala]: Thanks. I will make this change.

File: src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala (hunk 3 above)
Comment (reinierl) [inline src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala]: And I still don't see how it is a functional bug. The ChannelActor always processes each ChannelMessage with a valid channel that will *not* be closed while the operation is performed on the channel.

There is a performance penalty for renewing the channel twice instead of once of course, but indeed, I see no way to avoid that altogether.

Comment (wpc009) [inline src/main/scala/com/thenewmotion/akka/rabbitmq/ConnectionActor.scala]: I post some logs in the issue report. Maybe you could find some useful info there.

The problem I'm facing it real. I can easily reproduce it. The close operation on the extra channel cause some unexpected exception.

After this fix. the problem gone.

**Comment by jschaul** [comment]
Thanks @wpc009 for this PR, we're having similar issues with akka-rabbitmq version 2.2 and version 2.3 - upon a disconnect, reconnection fails in some cases (connection itself succeeds, but binding a consumer to a queue fails). Using the code from this PR no longer fails in these cases. 
Can we get the failing tests fixed and get this merged for a next release?
